strictdoc/export/excel/excel_generator.py

```diff
@@ -1,59 +1,193 @@
+
+        Returns:
+            int: Width of the Parent column
+        """
+        for column in EXPORT_COLUMNS:
+            column[MAX_WIDTH_KEY] = (
+                len(column[COLUMN_HEADER_KEY]) + HEADER_MARGIN
+            )
+        parent_header_len = len(PARENT_COLUMN_HEADER_LABEL) + HEADER_MARGIN
+
+        return parent_header_len
+
+    @staticmethod
+    def _set_columns_width(
+        workbook: Workbook, worksheet: Worksheet, parent_header_len: int
+    ):
+        """Set the width of each columns based on the max lenght and the max possible
+
+        Args:
+            workbook (Workbook): Excel workbook
+            worksheet (Worksheet): Excel worksheet
+            parent_header_len (int): Len of the parent column
+        """
+        cell_format_text_wrap = workbook.add_format()
+        cell_format_text_wrap.set_text_wrap()
+
+        for idx, column in enumerate(EXPORT_COLUMNS, start=0):
+            if column[MAX_WIDTH_KEY] > MAX_WIDTH:
+                worksheet.set_column(idx, idx, MAX_WIDTH, cell_format_text_wrap)
+            else:
+                worksheet.set_column(idx, idx, column[MAX_WIDTH_KEY])
+        worksheet.set_column(
+            len(EXPORT_COLUMNS), len(EXPORT_COLUMNS), parent_header_len
+        )
+
+    @staticmethod
+    def _init_headers() -> List:
+        """Init the header list
+
+        Returns:
+            List: List of header
+        """
+        headers: List = []
+
+        for column in EXPORT_COLUMNS:
+            headers.append({"header": column[COLUMN_HEADER_KEY]})
+        headers.append({"header": PARENT_COLUMN_HEADER_LABEL})
+
+        return headers
+
     @staticmethod
     def export_tree(
         document_tree: DocumentTree, traceability_index, output_excel_root
     ):
+        """Will create and fill one excel file per document, containing requierements with an UID
+
+        Args:
+            document_tree (DocumentTree): List of document to export
+            traceability_index (Any): traceability)index
+            output_excel_root (str): Path of the export
+        """
         Path(output_excel_root).mkdir(parents=True, exist_ok=True)
 
         document: Document
         for document in document_tree.document_list:
-            document_out_file = "{}/{}.xlsx".format(
-                output_excel_root, document.meta.document_filename_base
+            document_out_file = Path(
+                output_excel_root,
+                f"{document.meta.document_filename_base}.xlsx",
             )
 
-            workbook = xlsxwriter.Workbook(document_out_file)
-            worksheet = workbook.add_worksheet()
+            document_contents = list(
+                DocumentCachingIterator(document).all_content()
+            )  # not ideal but needed to make the first pass to detect refs
 
-            style_bold = workbook.add_format({'bold': True})
+            with xlsxwriter.Workbook(document_out_file) as workbook:
```

Comment:
This is the most important comment: I am getting the following crash trace here:

```
Traceback (most recent call last):
  File "/Users/Stanislaw/workspace/projects-code/strictdoc/strictdoc/cli/main.py", line 63, in <module>
    main()
  File "/Users/Stanislaw/workspace/projects-code/strictdoc/strictdoc/cli/main.py", line 55, in main
    args.input_paths, args.output_dir, args.formats
  File "/Users/Stanislaw/workspace/projects-code/strictdoc/strictdoc/helpers/timing.py", line 11, in wrap
    result = f(*args, **kw)
  File "/Users/Stanislaw/workspace/projects-code/strictdoc/strictdoc/core/actions/export_action.py", line 74, in export
    document_tree, traceability_index, output_excel_root
  File "/Users/Stanislaw/workspace/projects-code/strictdoc/strictdoc/export/excel/excel_generator.py", line 192, in export_tree
    workbook, worksheet, parent_header_len
  File "/Users/Stanislaw/.pyenv/versions/3.6.0/lib/python3.6/site-packages/xlsxwriter/workbook.py", line 168, in __exit__
    self.close()
  File "/Users/Stanislaw/.pyenv/versions/3.6.0/lib/python3.6/site-packages/xlsxwriter/workbook.py", line 320, in close
    self._store_workbook()
  File "/Users/Stanislaw/.pyenv/versions/3.6.0/lib/python3.6/site-packages/xlsxwriter/workbook.py", line 716, in _store_workbook
    xlsx_file.write(os_filename, xml_filename)
  File "/Users/Stanislaw/.pyenv/versions/3.6.0/lib/python3.6/zipfile.py", line 1608, in write
    with open(filename, "rb") as src, self.open(zinfo, 'w') as dest:
  File "/Users/Stanislaw/.pyenv/versions/3.6.0/lib/python3.6/zipfile.py", line 1348, in open
    return self._open_to_write(zinfo, force_zip64=force_zip64)
  File "/Users/Stanislaw/.pyenv/versions/3.6.0/lib/python3.6/zipfile.py", line 1461, in _open_to_write
    self.fp.write(zinfo.FileHeader(zip64))
  File "/Users/Stanislaw/.pyenv/versions/3.6.0/lib/python3.6/zipfile.py", line 721, in write
    n = self.fp.write(data)
AttributeError: 'PosixPath' object has no attribute 'write'
Exception ignored in: <bound method ZipFile.__del__ of <zipfile.ZipFile [closed]>>
Traceback (most recent call last):
  File "/Users/Stanislaw/.pyenv/versions/3.6.0/lib/python3.6/zipfile.py", line 1649, in __del__
    self.close()
  File "/Users/Stanislaw/.pyenv/versions/3.6.0/lib/python3.6/zipfile.py", line 1667, in close
    self._write_end_record()
  File "/Users/Stanislaw/.pyenv/versions/3.6.0/lib/python3.6/zipfile.py", line 1769, in _write_end_record
    self.fp.write(endrec)
  File "/Users/Stanislaw/.pyenv/versions/3.6.0/lib/python3.6/zipfile.py", line 721, in write
    n = self.fp.write(data)
AttributeError: 'PosixPath' object has no attribute 'write'
```

I am using Python 3.6.2 which is the current StrictDoc lowest version. Could check it on your side as well? Can it be that the `with ...` version of the open/close method is newer than 3.6.2?

Reply:
tested with python 3.6.2, no problem.
What OS? I am on Mac.

Reply:
I reverted to the previous implem for the path composition for now

Reply:
My apologies: I was testing it using 3.6.0. 3.6.2 works. I am also on Mac.

I am having another warning now when I run your branch on my own .sdoc doc set:

> /Users/Stanislaw/.pyenv/versions/3.6.2/lib/python3.6/site-packages/xlsxwriter/worksheet.py:2590: UserWarning: Must have at least one data row in in add_table()
  warn("Must have at least one data row in in add_table()")

I will check now what's happening.

Reply:
was able to replicate: you don't have any requirement with UID -> I only export the one with UID.
we need to decide what is the expected behavior in this case.

Reply:
Good catch! Let's send a warning early and do not try to generate Excel in this case.

Reply:
done